### DISCUSSION TITLE
How can I pass a variable to a json.stringify field?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#51064] [Help] [answered]
### Summary

Yes, I have been trying not to post and figured it out.... trying different approaches but I still have the same problems passing variables that I want it be reading as a string..

This time I want to do something like this... but it isn't working... 

 ```
body: JSON.stringify({
                    message: { getText },

```

This again is working on a Twitter API, and again it isn't liking the variables ...

complete code here:

```
import React from "react";
import { useEffect, useState } from "react";

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { cron } from "cron";

import Text from "../../components/text";
import db from "../../utils/text.json";

export async function getStaticProps() {
  return {
    props: { db },
  };
}

export default function TweetTest({ db }) {
  const [current1, setCurrent1] = useState("I am the current 1");

  useEffect(() => {
    console.log({ current1 });
  }, [current1]);

  const getText = () => {
    return db // Modified here
      .filter((item) => item.id === "1")
      .map((item) => <Text key={item.id} text={item.text} />);
  };

  const theString = getText.toString();

  return (
    <div className="main-about">
      <Container>
        <div> {} </div> <br />
        <div>
          Hello there {current1}
          <button
            type="button"
            className="btn btn-primary"
            onClick={async () => {
              try {
                await fetch("/api/twitterClient", {
                  method: "POST",
                  headers: {
                    "content-type": "application/json",
                  },
                  body: JSON.stringify({
                    message: { getText },
                  }),
                });
              } catch (e) {
                console.log(e); // fetch doesn't really throw on 4xx range
              }
            }}
          >
            Click
          </button>
        </div>
      </Container>
    </div>
  );
}

```

### Example

https://github.com/Elindo586/nextprojectd/blob/main/pages/twitter/test.js

**Comment by icyJoseph** [comment]
Hi,

You are not calling the `getText` function. That is `getText()`.

```jsx
 body: JSON.stringify({
                    message: getText() ,
                  }),
```

However, this seems strange to me because `getText`, when called `getText` returns a list of Text React Elements. It makes no sense to, or at least doesn't seem practical, to stringify that to send it in a POST request.

> You are also making this mistake in `const theString = getText.toString();`, you mean `const theString = getText()`. Remember, functions ought to be called!

I think want you want is to do find instead:

```jsx
  const getText = () => {
    return db.find((item) => item.id === "1")
  };
```

This returns the item with id, strictly equal to `"1"`, otherwise if the `db` list, has no such item, it returns undefined. Before using the result you should check if it exists.

Again, don't forget to call the function! `getText()`.

Also notice that using find will return the item, which is an object with an id and a text properties. You probably just want the text property.

**Comment by Elindo586** [comment]
So, I think I have tried this one before..
  ```
body: JSON.stringify({
                    message: getText() ,
                  }),
```

But it comes with this error from the API:
```
errors: [ [Object] ],
    title: 'Invalid Request',
    detail: 'One or more parameters to your request was invalid.',
    type: 'https://api.twitter.com/2/problems/invalid-request'
  },
  page: '/api/twitterClient',
  errors: [ { message: 'Request body must be a JSON object.' } ]

```
Almost feels as if Twitter is preventing this type of coding using variables, of course that might not be the case and might be I am just terrible at coding...


At this point is just trial and error without knowing how/why things work...

I also tried this one:
```
const theString = getText().toString();

 body: JSON.stringify({ message: theString() }),
```

It was able to tweet this below, and I don't know why it came out that way for example
`[object Object]`



I am thinking I might need to use `.parse` , but that means re-do most of the code, but that seems silly too because if the problem comes from the twitter API chances are I might end up with the same errors...

**Comment by icyJoseph** [comment]
@Elindo586 

The problem is that with your current state of code, `getText`  when called, returns a list of `<Text key={item.id} text={item.text} />` 

Which are all objects. These are called React elements, and when you call `.toString()` on it, it becomes `[object Object]`.

I think you are just juggling with many things at the same time. Eventually you'll develop intuition and tools of your own to develop this.

This is the code that'll work:

```js
import { useEffect, useState } from "react";

import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

import { cron } from "cron";

import Text from "../../components/text";
import db from "../../utils/text.json";

export async function getStaticProps() {
  return {
    props: { db },
  };
}

export default function TweetTest({ db }) {
  const [current1, setCurrent1] = useState("I am the current 1");

  useEffect(() => {
    console.log({ current1 });
  }, [current1]);

  const getText = () => {
    const target = db.find((item) => item.id === "1");

    if (!target) return "";

    return target.text;
  };

  return (
    <div className="main-about">
      <Container>
        <div> {} </div> <br />
        <div>
          Hello there {current1}
          <button
            type="button"
            className="btn btn-primary"
            onClick={async () => {
              try {
                await fetch("/api/twitterClient", {
                  method: "POST",
                  headers: {
                    "content-type": "application/json",
                  },
                  body: JSON.stringify({
                    message: getText(),
                  }),
                });
              } catch (e) {
                console.log(e); // fetch doesn't really throw on 4xx range
              }
            }}
          >
            Click
          </button>
        </div>
      </Container>
    </div>
  );
}
```

**Reply by icyJoseph** [reply, marked as answer]
A couple of things/tips and encouragement:

- While learning, your first debugging tool is `console.log`, you could've called `console.log(getText())` to see what's going on there
- Notice that the `getText` function now does a `find`, which returns the first element for which the `callback/predicate` is true
- `find` might not find anything at all, so it is a good idea to make a check, such as `if (!target)`
- Making assumptions is good, as long as, we also make a habit of trying to device counter examples. You believed that somehow the twitter API does not allow you to pass variables, or that maybe JSON.stringify did that.

This code change :

```js
           async () => {
              const message = "hello world";
              try {
                await fetch("/api/twitterClient", {
                  method: "POST",
                  headers: {
                    "content-type": "application/json",
                  },
                  body: JSON.stringify({
                    message: message,
                  }),
                });
              } catch (e) {
                console.log(e);
              }
            }
```

Would've proven the assumption to be incorrect.

- use `typeof `, for example `console.log(typeof getText)` versus `console.log(typeof getText())` - function and object, this can give you certainty as to what value type are you really working with.
- identify the actual issue, here the problem has been all along `getText`, and what happens when you call it. When faced with a situation that you feel is going over your skill level, simplify, remove parts, until there's only the thing you are actually trying to do left. For example, calling `getText`, and getting the string your need. 

Learning to use APIs/SDKs, is the kind of experience that can get frustrating very quickly. This is a good time to develop your own tools. Use `console.log`, both in the client and server, look at the browser DevTools, the network tab is helpful here. Maybe use a `debugger` statement and step through your code, ultimately this is the best tool.

I think it is also time to trust yourself a bit more too.

**Reply by Elindo586** [reply]
Well thanks... it worked...

Yes, the API was telling me "object" and they were expecting something different... so I went to test a similar setup that didn't involve the API wrapper, and the test did work (here https://github.com/Elindo586/nextprojectd/blob/main/pages/test.js) , but obviously I was using it in a different way that didn't give me that error...

And yes... I had the feeling that I needed to pin point the .text field, and I had no idea it was returning an object even when the error was telling me it was an object..
